runtime/account_storage_v1.go

```diff
@@ -0,0 +1,218 @@
+		s.ledger,
+		s.slabStorage,
+		address,
+		domain,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	if domainStorageMap == nil && createIfNotExists {
+		domainStorageMap = s.storeNewDomainStorageMap(address, domain)
+	}
+
+	return domainStorageMap
+}
+
+func (s *AccountStorageV1) storeNewDomainStorageMap(
+	address common.Address,
+	domain common.StorageDomain,
+) *interpreter.DomainStorageMap {
+
+	domainStorageMap := interpreter.NewDomainStorageMap(
+		s.memoryGauge,
+		s.slabStorage,
+		atree.Address(address),
+	)
+
+	slabIndex := domainStorageMap.SlabID().Index()
+
+	storageKey := interpreter.NewStorageDomainKey(s.memoryGauge, address, domain)
+
+	if s.newDomainStorageMapSlabIndices == nil {
+		s.newDomainStorageMapSlabIndices = map[interpreter.StorageDomainKey]atree.SlabIndex{}
+	}
+	s.newDomainStorageMapSlabIndices[storageKey] = slabIndex
+
+	return domainStorageMap
+}
+
+func (s *AccountStorageV1) commit() error {
+
+	switch len(s.newDomainStorageMapSlabIndices) {
+	case 0:
+		// Nothing to commit.
+		return nil
+
+	case 1:
+		// Optimize for the common case of a single domain storage map.
+
+		var updated int
+		for storageDomainKey, slabIndex := range s.newDomainStorageMapSlabIndices { //nolint:maprange
+			if updated > 0 {
+				panic(errors.NewUnreachableError())
+			}
+
+			err := s.writeStorageDomainSlabIndex(
+				storageDomainKey,
+				slabIndex,
+			)
+			if err != nil {
+				return err
+			}
+
+			updated++
+		}
+
+	default:
+		// Sort the indices to ensure deterministic order
+
+		type domainStorageMapSlabIndex struct {
+			StorageDomainKey interpreter.StorageDomainKey
+			SlabIndex        atree.SlabIndex
+		}
+
+		slabIndices := make([]domainStorageMapSlabIndex, 0, len(s.newDomainStorageMapSlabIndices))
+		for storageDomainKey, slabIndex := range s.newDomainStorageMapSlabIndices { //nolint:maprange
+			slabIndices = append(
+				slabIndices,
+				domainStorageMapSlabIndex{
+					StorageDomainKey: storageDomainKey,
+					SlabIndex:        slabIndex,
+				},
+			)
+		}
+		sort.Slice(
+			slabIndices,
+			func(i, j int) bool {
+				slabIndex1 := slabIndices[i]
+				slabIndex2 := slabIndices[j]
+				domainKey1 := slabIndex1.StorageDomainKey
+				domainKey2 := slabIndex2.StorageDomainKey
+				return domainKey1.Compare(domainKey2) < 0
+			},
+		)
+
+		for _, slabIndex := range slabIndices {
+			err := s.writeStorageDomainSlabIndex(
+				slabIndex.StorageDomainKey,
+				slabIndex.SlabIndex,
+			)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	s.newDomainStorageMapSlabIndices = nil
+
+	return nil
+}
+
+func (s *AccountStorageV1) writeStorageDomainSlabIndex(
+	storageDomainKey interpreter.StorageDomainKey,
+	slabIndex atree.SlabIndex,
+) error {
+	var err error
+	errors.WrapPanic(func() {
+		err = s.ledger.SetValue(
+			storageDomainKey.Address[:],
+			[]byte(storageDomainKey.Domain.Identifier()),
+			slabIndex[:],
+		)
+	})
+	if err != nil {
+		return interpreter.WrappedExternalError(err)
+	}
+	return nil
+}
```

Comment:
Maybe refactor both `AccountStorageV1.writeStorageDomainSlabIndex()` and `AccountStorageV2.writeAccountStorageSlabIndex()` by having a function to write `atree.SlabIndex` to register.  

Since the new function would be counterpart of existing `getSlabIndexFromRegisterValue()`, maybe also move these register value reading and writing functions together to another file.

Reply:
Great ideas! 

> Maybe refactor both `AccountStorageV1.writeStorageDomainSlabIndex()` and `AccountStorageV2.writeAccountStorageSlabIndex()` by having a function to write `atree.SlabIndex` to register.

Done in [4addf32](https://github.com/onflow/cadence/pull/3678/commits/4addf32051aa9cf9f39e18a3d4036e6e9af1d315)


> Since the new function would be counterpart of existing `getSlabIndexFromRegisterValue()`, maybe also move these register value reading and writing functions together to another file.



Done in [4fd610a](https://github.com/onflow/cadence/pull/3678/commits/4fd610ac529be171c69c20ca024ef7e76af09d1e)